src/Integration/LocalServices/UnboundProjectFinder.cs

```diff
@@ -76,31 +84,80 @@ public IEnumerable<Project> GetUnboundProjects()
 
         private Project[] GetUnboundProjects(BindingConfiguration binding)
         {
+            threadHandling.ThrowIfOnUIThread();
+
             Debug.Assert(binding.Mode.IsInAConnectedMode());
             Debug.Assert(binding.Project != null);
 
-            var filteredSolutionProjects = projectSystem.GetFilteredSolutionProjects().ToArray();
-            logger.LogDebug($"[Binding check] Number of bindable projects: {filteredSolutionProjects.Length}");
+            // Using threadHandling.Run(...) here to avoid having to convert all of the upstream
+            // // calling methods to be async.
+            var projects = threadHandling.Run(() => GetUnboundProjectsAsync(binding));
+
+            threadHandling.ThrowIfOnUIThread();
+            return projects;
+        }
 
-            var result = filteredSolutionProjects
-                .Where(project =>
+        private async Task<Project[]> GetUnboundProjectsAsync(BindingConfiguration binding)
+        {
+            threadHandling.ThrowIfOnUIThread();
+
+            Debug.Assert(binding.Mode.IsInAConnectedMode());
+            Debug.Assert(binding.Project != null);
+
+            var filteredSolutionProjects = await GetFilteredSolutionProjectsAsync();
+            Debug.Assert(!threadHandling.CheckAccess(), "Should not be on the UI thread");
+
+            var unboundProjects = new List<Project>();
+            foreach (var project in filteredSolutionProjects)
+            {
+                var isUnbound = await IsBindingRequiredAsync(project, binding);
+
+                if (isUnbound)
                 {
-                    var configProjectBinder = projectBinderFactory.Get(project);
+                    unboundProjects.Add(project);
+                }
+            }
+
+            return unboundProjects.ToArray();
+        }
+
+        private async Task<Project[]> GetFilteredSolutionProjectsAsync()
+        {
+            threadHandling.ThrowIfOnUIThread();
```

Comment:
Expecting to start on a background thread, but switches to the UI thread to do the work, blocking the background thread until it is done.